tiledb/sm/query/query_condition.cc

```diff
@@ -671,46 +687,158 @@ QueryCondition::apply_clause(
       return {Status_QueryConditionError(
                   "Cannot perform query comparison; Unsupported query "
                   "conditional type on " +
-                  clause.field_name_),
+                  node->field_name_),
               nullopt};
   }
 
   return {Status::Ok(), nullopt};
 }
 
+static std::vector<ResultCellSlab> result_slabs_merge_union(
+    const std::vector<ResultCellSlab>& arr1,
+    const std::vector<ResultCellSlab>& arr2) {
+  size_t iter1 = 0;
+  size_t iter2 = 0;
+  std::vector<ResultCellSlab> ret;
+  while (iter1 < arr1.size() && iter2 < arr2.size()) {
+    ResultCellSlab less;
+    if (arr1[iter1].start_ < arr2[iter2].start_) {
+      less = arr1[iter1];
+      iter1 += 1;
+    } else {
+      less = arr2[iter2];
+      iter2 += 1;
+    }
+
+    if (ret.size() > 0 &&
+        ret.back().start_ + ret.back().length_ >= less.start_) {
+      ret.back().length_ = less.start_ + less.length_ - ret.back().start_;
+    } else {
+      ret.push_back(less);
+    }
+  }
+
+  while (iter1 < arr1.size()) {
+    ret.push_back(arr1[iter1]);
+    iter1 += 1;
+  }
+
+  while (iter2 < arr2.size()) {
+    ret.push_back(arr2[iter2]);
+    iter2 += 1;
+  }
+  return ret;
+}
+
+static std::vector<ResultCellSlab> result_slabs_merge_intersection(
+    const std::vector<ResultCellSlab>& arr1,
+    const std::vector<ResultCellSlab>& arr2) {
+  size_t iter1 = 0;
+  size_t iter2 = 0;
+  std::vector<ResultCellSlab> ret;
+  while (iter1 < arr1.size() && iter2 < arr2.size()) {
+    // interval is [start, end)
+    size_t start = std::max(arr1[iter1].start_, arr2[iter2].start_);
+    size_t end = std::min(
+        arr1[iter1].start_ + arr1[iter1].length_,
+        arr2[iter2].start_ + arr2[iter2].length_);
+
+    if (start < end) {
+      ResultCellSlab c(arr1[iter1].tile_, start, end - start);
+      ret.push_back(c);
+    }
+
+    if (arr1[iter1].start_ + arr1[iter1].length_ <
+        arr2[iter2].start_ + arr2[iter2].length_) {
+      iter1 += 1;
+    } else {
+      iter2 += 1;
+    }
+  }
+
+  return ret;
+}
+
+std::vector<ResultCellSlab> QueryCondition::apply_tree(
+    const tdb_unique_ptr<ASTNode>& node,
+    const ArraySchema& array_schema,
+    uint64_t stride,
+    const std::vector<ResultCellSlab>& result_cell_slabs) const {
+  // how to implement?
```

Comment:
Adjust comment?

Reply:
Actually, I'm not sure how this method is supposed to work. result_slabs_merge_intersection and result_slabs_merge_union have no function documentation or code comments. apply_tree should also include more code comments.

Reply:
Resolved locally